helix-core/src/text_annotations.rs

```diff
@@ -77,45 +79,122 @@ impl Overlay {
     }
 }
 
-/// Line annotations allow for virtual text between normal
-/// text lines. They cause `height` empty lines to be inserted
-/// below the document line that contains `anchor_char_idx`.
-///
-/// These lines can be filled with text in the rendering code
+/// Line annotations allow inserting virtual text lines between normal
+/// text lines.  These lines can be filled with text in the rendering code
 /// as their contents have no effect beyond visual appearance.
 ///
-/// To insert a line after a document line simply set
-/// `anchor_char_idx` to `doc.line_to_char(line_idx)`
-#[derive(Debug, Clone)]
-pub struct LineAnnotation {
-    pub anchor_char_idx: usize,
-    pub height: usize,
+/// The height of virtual text is usually not know ahead of time
+/// as virtual text often requires softwrapping. Furthermore the
+/// height of some virtual text like some virtual text like side-by-side
+/// diffs depends on the height of the text (again influcend by softwrap)
+/// and other virtual text  Therefore line annotations are computed
+/// on the fly instead of ahead of time like other annotations.
+///
+/// The core of this trait `insert_virtual_lines` function. It is called at
+/// the end of every  visual line and allows the `LineAnnotation` to insert
+/// empty virtual lines. Apart from that the `LineAnnotation` trait has
+/// multiple methods that allow it to track anchors in the document.
+///
+/// When a new transerversal of a document starts `reset_pos` is called.
+/// Afterwards the other functions are called with indecies that are larger
+/// then the one passed to `reset_pos`. This allows performing a binary search (use `partition_point`)
+/// in `reset_pos` once and then to only look at the next anchor during each
+/// method call.
+///
+/// The `reset_pos`, `skip_conceal` and `process_anchor` functions return a `char_idx`
+/// anchor. This anchor is stored when transvering the document and when the grapheme at
+/// the anchor is transvered the `process_anchor` function is called.
+///
+/// # Note
+///
+/// All functions only recive immutable reference to `self`.
+/// LineAnnotations that want to store an internal position or
+/// state of somekind should use `Cell`. Using interior mutability for
+/// caches is actually preffered as otherwise a lot of lieftimes would
+/// become invariant which can cause weird lifetime issues and unwieldy API.
```

Comment:
```suggestion
/// All functions only receive immutable references to `self`.
/// `LineAnnotation`s that want to store an internal position or
/// state of some kind should use `Cell`. Using interior mutability for
/// caches is actually preferred as otherwise a lot of lifetimes would
/// become invariant which can cause weird lifetime issues and unwieldy APIs.
```